File: OpenLab.Kitchen.Receiver.RFID/FeigRFID/Samples/VB/HexEditor/VS2010/HexEditor.vb
'---------------------------------------------------------------
'Copyright © 2003-2008	FEIG ELECTRONIC GmbH, All Rights Reserved.
'						Lange Strasse 4
'						D-35781 Weilburg
'						Federal Republic of Germany
'						phone    : +49 6471 31090
'						fax      : +49 6471 310999
'						e-mail   : obid-support@feig.de
'						Internet : http://www.feig.de
'
'OBID® and OBID i-scan® are registered trademarks of FEIG ELECTRONIC GmbH
'----------------------------------------------------------------

Imports System

Namespace OBID

    Public Class HexEditor
        Inherits System.Windows.Forms.UserControl

#Region " Vom Windows Form Designer generierter Code "

        Public Sub New()
            MyBase.New()

            ' Dieser Aufruf ist für den Windows Form-Designer erforderlich.
            InitializeComponent()

            ' Initialisierungen nach dem Aufruf InitializeComponent() hinzufügen

        End Sub

        ' UserControl1 überschreibt den Löschvorgang zur Bereinigung der Komponentenliste.
        Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If Not (components Is Nothing) Then
                    components.Dispose()
                End If
            End If
            MyBase.Dispose(disposing)
        End Sub

        ' Für Windows Form-Designer erforderlich
        Private components As System.ComponentModel.IContainer

        'HINWEIS: Die folgende Prozedur ist für den Windows Form-Designer erforderlich
        'Sie kann mit dem Windows Form-Designer modifiziert werden.
        'Verwenden Sie nicht den Code-Editor zur Bearbeitung.
        Friend WithEvents Hexer As System.Windows.Forms.TextBox
        <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
            Me.Hexer = New System.Windows.Forms.TextBox
            Me.SuspendLayout()
            '
            'Hexer
            '
            Me.Hexer.Anchor = CType((((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Bottom) _
                        Or System.Windows.Forms.AnchorStyles.Left) _
                        Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
            Me.Hexer.Font = New System.Drawing.Font("Courier New", 8.25!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.Hexer.Location = New System.Drawing.Point(0, 0)
            Me.Hexer.Multiline = True
            Me.Hexer.Name = "Hexer"
            Me.Hexer.ScrollBars = System.Windows.Forms.ScrollBars.Both
            Me.Hexer.Size = New System.Drawing.Size(272, 256)
            Me.Hexer.TabIndex = 1
            Me.Hexer.Text = ""
            Me.Hexer.WordWrap = False
            '
            'UserControl1
            '
            Me.Controls.Add(Me.Hexer)
            Me.Name = "UserControl1"
            Me.Size = New System.Drawing.Size(272, 256)
            Me.ResumeLayout(False)

        End Sub

#End Region
        Public Sub InsertData(ByVal index As Integer, ByVal b As Byte())
            If b Is Nothing Then Return

            Dim s As String
            Dim i As Integer

            row = Fix(index / _columnCount)
            column = (index Mod _columnCount) * 3 + 8
            For i = 0 To b.Length - 1
                s = Trim(b(i).ToString("X"))
                If s.Length = 1 Then s = "0" + s
                InsertText(Trim(Mid(s, 1)))
                InsertText(Trim(Mid(s, 2)))
            Next
        End Sub

        Public Function GetData(ByVal index As Integer, ByVal count As Integer) As Byte()
            Dim b(count) As Byte
            Dim i As Integer
            Dim s As String

            For i = 0 To count - 1
                s = GetText(index + i)
                If s = "" Then Exit Function
                b(i) = OBID.FeHexConvert.HexStringToByte(s)
            Next

            Return b
        End Function

        Private Sub Hexer_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles Hexer.MouseUp
            Hexer.SelectionLength = 0
            row = Fix(Hexer.SelectionStart / (_columnCount * 4 + 10))
            column = Hexer.SelectionStart Mod (_columnCount * 4 + 10)
            ' Cursor auf die richtige Position setzen fall in dem Hex-Bereich geklickt wurde
            If column < 8 Then column = 8
            If column < 7 + _columnCount * 3 Then
                If (column - 7) Mod 3 = 0 Then
                    column = column + 1
                End If
            End If
            Hexer.SelectionStart = row * (_columnCount * 4 + 10) + column
        End Sub

        Private Sub Hexer_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles Hexer.KeyPress
            Select Case e.KeyChar
                Case "0" To "9", "a" To "f", "A" To "F"
                    InsertText(e.KeyChar.ToString)
                    e.Handled = True
                Case Else
                    e.Handled = True
            End Select

        End Sub

        Private Sub InsertText(ByVal s As String)
            If (column > 7) And (column < 7 + _columnCount * 3) Then
                Dim pos As Integer
                Dim tmp As String
                column = column + 1
                pos = row * (_columnCount * 4 + 10) + column
                If pos >= Hexer.Text.Length Then Exit Sub
                Mid(Hexer.Text, pos) = s.ToUpper

                ' Block herausnehmen
                If column Mod 3 = 0 Then
                    tmp = Mid(Hexer.Text, pos, 2)
                End If
                If column Mod 3 = 1 Then
                    tmp = Mid(Hexer.Text, pos - 1, 2)
                End If
                ' Position für die ASCII-Schreibweise berechnen
                pos = row * (_columnCount * 4 + 10) + 3 * _columnCount + 9 + Fix((column - 9) / 3)
                Dim i As Integer
                i = OBID.FeHexConvert.HexStringToByte(tmp)
                ' Es sollen nur druckbare Zeichen eingefügen
                Select Case i
                    Case 21 To 255
                        Mid(Hexer.Text, pos) = Chr(i)
                End Select

                ' nur zweierblöcke sollen eingegeben werden können
                If (column - 7) Mod 3 = 0 Then
                    column = column + 1
                End If
                ' Am Ende der Zeile soll der Cursor in die neue Zeile springen
                If column = 7 + _columnCount * 3 Or column = 8 + _columnCount * 3 Then
                    column = 8
                    row = row + 1
                End If
                ' Für die bessere Optik soll auch der Cursor an der richtigen Position angezeigt werden
                Hexer.SelectionLength = 0
                Hexer.SelectionStart = row * (_columnCount * 4 + 10) + column
            End If
        End Sub

        Private Function GetText(ByVal index As Integer) As String
            row = Fix(index / _columnCount)
            column = (index Mod _columnCount) * 3 + 9

            Dim pos As Integer
            Dim s As String
            pos = row * (_columnCount * 4 + 10) + column
            If pos >= Hexer.Text.Length - 5 Then Exit Function
            s = Mid(Hexer.Text, pos, 2)

            Return s
        End Function

        Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
            SetSize(128, 4)
        End Sub

        Public Sub SetSize(ByVal rowcount As Integer, ByVal columncount As Integer)
            Dim i As Integer
            Dim j As Integer

            If _rowCount = rowcount And _columnCount = columncount Then
                Return
            End If

            _rowCount = rowcount
            _columnCount = columncount
            Hexer.Visible = False

            Hexer.Text = ""
            For i = 0 To rowcount - 1
                Hexer.AppendText("DB ")
                If i < 10 Then
                    Hexer.AppendText(" ")
                End If
                If i < 100 Then
                    Hexer.AppendText(" ")
                End If
                Hexer.AppendText(i.ToString + ": ")
                For j = 0 To columncount - 1
                    Hexer.AppendText("00 ")
                Next
                For j = 0 To columncount - 1
                    Hexer.AppendText(".")
                Next
                Hexer.AppendText(vbNewLine)
            Next
            Hexer.Visible = True
        End Sub

        Public Property ColumnCount() As Integer
            Get
                Return _columnCount
            End Get
            Set(ByVal Value As Integer)
                Dim b() As Byte

                b = Me.GetData(0, _columnCount * _rowCount)
                Me.SetSize(_rowCount, Value)
                Me.InsertData(0, b)
            End Set
        End Property

        Public ReadOnly Property RowCount() As Integer
            Get
                Return _rowCount
            End Get
        End Property

        Private row As Integer
        Private column As Integer
        Private _columnCount As Integer
        Private _rowCount As Integer
    End Class
End Namespace

File: OpenLab.Kitchen.Receiver.RFID/FeigRFID/Samples/C#/ISOHostSample/VS2010/HexEditor.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;

namespace ISOHostSample
{
    public partial class HexEditor : UserControl
    {
        private int row;
        private int column;
        private int _columnCount;
        private int _rowCount;
        private int bnkidx;

        public HexEditor()
        
        {
            InitializeComponent();
            
        }

        public void InsertData(int index, byte[] b)
        {

            if (b == null)
            { return;
            }
            else
            {
                string s;
                int i;
                //just the interger number
                row = Convert.ToInt16(index / _columnCount);
                column = (index % _columnCount) * 3 + 8;
                for (i = 0; i <= b.Length - 1; i++)
                {
                    s = (b[i].ToString("X")).Trim();
                    if (s.Length == 1)
                    { s = "0" + s; }
                    //just insert this two string
                    InsertText(s.Substring(0).Trim());
                    InsertText(s.Substring(1).Trim());
                }
                
            }
        }
        
        public byte[] GetData(int index, int count)
        {
            byte[] b = new byte[count];
            int i;
            string s;

            for (i = 0; i <count; i++)
                
            {
                s = GetText(index,i);
                b[i] = OBID.FeHexConvert.HexStringToByte(s);
            }
            return b;
        }
                    
        private void InsertText(string s)
        {
            if ((column > 7) && (column < 7 + _columnCount * 3))
            {
                int pos;
                string tmp = null;
                column = column + 1;
                pos = row * (_columnCount * 4 + 10) + column;

                if (pos >= Hexer.Text.Length)
                    return; 
                
                //remove and insert the new byte
                Hexer.Text = Hexer.Text.Remove(pos - 1, s.Length);
                Hexer.Text = Hexer.Text.Insert(pos - 1, s.ToUpper());

                
                //write each byte on each column
                if (column % 3 == 0)
                {
                    tmp = Hexer.Text.Substring(pos - 1, 2);
                }
                if (column % 3 == 1)
                {
                    tmp = Hexer.Text.Substring(pos - 2, 2);
                }
                // Position for Ascii corversion
                pos = row * (_columnCount * 4 + 10) + 3 * _columnCount + 9 + Convert.ToInt16((column - 9) / 3);
                byte i;
                i = OBID.FeHexConvert.HexStringToByte(tmp); 

                //just regular number or letter
                if (i >= 21 && i <= 255)
                {
                    Hexer.Text = Hexer.Text.Remove(pos - 1, 1);
                    Hexer.Text = Hexer.Text.Insert(pos - 1, Convert.ToChar(i).ToString()); 
                }
                // next column
                if ((column - 7) % 3 == 0)
                {
                    column = column + 1;
                }
                // next line
                if (column == 7 + _columnCount * 3 || column == 8 + _columnCount * 3)
                {
                    column = 8;
                    row = row + 1;
                }
                //always put the cursor in the correct position 
                Hexer.SelectionLength = 0;
                Hexer.SelectionStart = row * (_columnCount * 4 + 10) + column;
            }
            else 
            {
                MessageBox.Show("This Fied is just lisible not writable");
            }
        }

        private string GetText(int index,int i )
        {
            int pos;
            pos = 3 * i + _columnCount * index + 13 * index + 3 * (_columnCount - 1)*index + 9;
            string s;
            if (pos >= Hexer.Text.Length - 6)
            { 
                return null ; 
            }   
            s = Hexer.Text.Substring(pos-1, 2);

            return s;
        }
        
        public void SetSize(int rowcount, int columncount, int Bnkidx)
        {
            int i;
            int j;
            string IniText = "";
            bnkidx = Bnkidx;
            

            _rowCount = rowcount;
            _columnCount = columncount;
           
            
            Hexer.Text = "";

            for (i = 0; i <= _rowCount - 1; i++)
            {
                switch (bnkidx)
                {
                    case 0:
                        IniText += "RES";
                        break;
                    case 1:
                        IniText += "EPC";
                        break;
                    case 2:
                        IniText += "TID";
                        break;
                    case 3:
                        IniText += "DB ";
                        break;
                    default:
                        return;
                }               
                if (i < 10)
                {
                    IniText +=  " ";
                }
                if (i < 100)
                {
                    IniText +=  " ";
                }
                IniText = IniText + i.ToString() + ": ";
                for (j = 0; j <= _columnCount - 1; j++)
                {
                    IniText +=  "00 ";
                }
                for (j = 0; j <= _columnCount - 1; j++)
                {
                    IniText +=  ".";
                }
                IniText += Environment.NewLine;
            }
              
              Hexer.AppendText(IniText) ;
      }

        public int ColumnCount
        {
            get
            {
                return _columnCount;
            }
            
            set
            {
                byte[] b;
                

                b = this.GetData(0, _columnCount * _rowCount);
                this.SetSize(_rowCount, value,  bnkidx);
                this.InsertData(0, b);
               
            }
        }

        public  int RowCount
        {
            get
            {
                return _rowCount;
            }
        }
         
        //Clean the HexEditor
        public void HexEditClean(int rowcount, int columncount, int Bnkidx)
        {
            _rowCount = rowcount;
            _columnCount = columncount;
            bnkidx = Bnkidx;

            int i;
            int j;
            string Text = "";

            Hexer.Text = "";
            for (i =0 ; i <=_rowCount  - 1; i++)
            {

                switch (bnkidx)
                {
                    case 0:
                        Text += "RES";
                        break;
                    case 1:
                        Text += "EPC";
                        break;
                    case 2:
                        Text += "TID";
                        break;
                    case 3:
                        Text += "DB ";
                        break;
                    default:
                        return;
                }
                if (i < 10)
                {
                    Text += " ";
                }
                if (i < 100)
                {
                    Text +=  " ";
                }
                Text = Text + i.ToString() + ": ";
                for (j = 0; j <= _columnCount - 1; j++)
                {
                    Text += "00 ";
                }
                    for (j = 0; j <= _columnCount - 1; j++)
                    {
                        Text +=  ".";
                    }
                    Text += Environment.NewLine; 
            }
            Hexer.AppendText(Text);
           
        }

        private void Hexer_MouseUp(object sender, MouseEventArgs e)
        {
            this.Hexer.SelectionLength = 0;
            row = Convert.ToInt16((Hexer.SelectionStart / (_columnCount * 4 + 10)));//Fix
            column = Hexer.SelectionStart % (_columnCount * 4 + 10);
            // Cursor auf die richtige Position setzen fall in dem Hex-Bereich geklickt wurde
            if (column < 8) { column = 8; }
            if (column < 7 + _columnCount * 3)
            {
                if ((column - 7) % 3 == 0)
                {
                    column = column + 1;
                }
            }
            Hexer.SelectionStart = row * (_columnCount * 4 + 10) + column;

        }

        private void Hexer_KeyPress_1(object sender, KeyPressEventArgs e)
        {
            
            //Ascii  dezcimale Value
           int dec = e.KeyChar;
                   
            if ((dec >= 48 && dec <= 57) || (dec >= 97 && dec <= 102) || (dec >= 65 && dec <= 70))
            {
                
                InsertText(e.KeyChar.ToString());
                e.Handled = true;
                return;
            }
            else
            {
                MessageBox.Show("just give a character in range (a-f) or (A-F) or(0-9)");
                e.Handled = true;
                return;
            }

        }
}
}

